wasmplugin/plugin.go

```diff
@@ -187,41 +191,63 @@ func (ctx *httpContext) OnHttpRequestBody(bodySize int, endOfStream bool) types.
 	// Do not perform any action related to request body if SecRequestBodyAccess is set to false
 	if !tx.IsRequestBodyAccessible() {
 		proxywasm.LogDebug("skipping request body inspection, SecRequestBodyAccess is off.")
+		ctx.processedRequestBody = true
+		interruption, err := tx.ProcessRequestBody()
+		if err != nil {
+			proxywasm.LogCriticalf("failed to process request body: %v", err)
+			return types.ActionContinue
+		}
+
+		if interruption != nil {
+			return ctx.handleInterruption("http_request_body", interruption)
+		}
+
 		return types.ActionContinue
 	}
 
-	ctx.requestBodySize += bodySize
-	// Wait until we see the entire body. It has to be buffered in order to check that it is fully legit
-	// before sending it upstream
-	if !endOfStream {
-		return types.ActionPause
-	}
+	if bodySize > 0 {
+		b, err := proxywasm.GetHttpRequestBody(ctx.bodyReadIndex, bodySize)
+		if err == nil {
+			interruption, _, err := tx.WriteRequestBody(b)
+			if err != nil {
+				proxywasm.LogCriticalf("failed to write request body: %v", err)
+				return types.ActionContinue
```

Comment:
The behaviour when an error is raised is something we may have to discuss further in another PR. `types.ActionContinue` sends everything upstream, it may not be the safest approach from a security standpoint. For example, receiving multiple chunks of the body with the latest that leads to an error. The whole body is sent upstream bypassing the WAF.
We may:
- Still try to run `ProcessX`, looking at interruption raised by processing the data that we already have
- Take a more conservative approach and drop any request that leads to an internal error